rezervo/sessions.py

```diff
@@ -93,3 +112,85 @@ async def remove_session(
             chain=chain_identifier, user_id=user_id, class_id=class_id
         ).delete()
         db.commit()
+
+
+async def remove_sessions(
+    chain_identifier: ChainIdentifier,
+    user_id: UUID,
+    recurrent_ids_to_remove: list[str],
+    session_state: SessionState,
+):
+    with SessionLocal() as db:
+        user_planned_sessions = (
+            db.query(models.Session)
+            .filter(
+                models.Session.chain == chain_identifier,
+                models.Session.user_id == user_id,
+                models.Session.status == session_state,
+            )
+            .all()
+        )
+        sessions_to_remove = [
+            session
+            for session in user_planned_sessions
+            if rezervo_class_recurrent_id(UserSession.from_orm(session).class_data)
+            in recurrent_ids_to_remove
+        ]
+        for session in sessions_to_remove:
+            db.delete(session)
+        db.commit()
+
+
+async def update_planned_sessions(
+    chain_identifier: ChainIdentifier,
+    user_id: UUID,
+    previous_config: ChainConfig | None,
+    updated_config: ChainConfig,
+):
+    previous_class_ids = {
+        class_config_recurrent_id(_class)
+        for _class in (previous_config.recurring_bookings if previous_config else [])
+    }
+    updated_class_ids = {
+        class_config_recurrent_id(_class)
+        for _class in updated_config.recurring_bookings
+    }
+
+    config_active_changed = (
+        previous_config is None or previous_config.active != updated_config.active
+    )
+    removed_class_ids = (
+        updated_class_ids
+        if config_active_changed and not updated_config.active
+        else previous_class_ids - updated_class_ids
+    )
+    added_class_ids = (
+        updated_class_ids
+        if config_active_changed and updated_config.active
+        else updated_class_ids - previous_class_ids
+    )
```

Comment:
There are some cases where this behaves incorrectly.

If the config stays inactive but classes are changed (`previous_config.active=false`,`updated_config.active=false` and `config_active_changed=false`), nothing should be done. However, now it will behave as if classes were changed while active, resulting in removing non-existent sessions and adding planned sessions even though booking is inactive.

Also, if you changed it from inactive to active, all classes from the new config should be added and nothing removed (since no sessions from the previous config exist, because it was not active). What actually happens is that we attempt to remove sessions for classes not in the new config.

I believe the below suggestion will perform the correct changes in all cases, but please double check :hugs:  
```suggestion
    added_class_ids: set[str] = updated_class_ids
    removed_class_ids: set[str] = set()
    if previous_config is not None and previous_config.active and updated_config.active:
        # common case: booking is active, just changing which classes to book
        # add all new arrivals and remove all departures
        added_class_ids -= previous_class_ids
        removed_class_ids = previous_class_ids - updated_class_ids
    elif not updated_config.active:
        # booking deactivated, remove any previously planned sessions (nothing if already inactive)
        added_class_ids = set()
        removed_class_ids = previous_class_ids
```

Reply:
Actually, scratch that suggestion. It is unnecessarily convoluted. You can simply say that the class id lists are empty if the respective config is either `None` or has `active=False`. That will reflect the actual existing and desired state of the sessions table. Then you can easily calculate the required additions and removals.

Reply:
Sure, that looks sound! https://github.com/mathiazom/rezervo/pull/46/commits/69a3e1641919e5a933ffbc864249b760419e3d81

Reply:
Works well in my testing also